File: ROID-Forum-Web-Client/Convert Client Code/GMGWebsite/web/Website/Sections/CodingSection/ThreadViews/ThreadDisplay.dart
library CodingThreadDisplay;
import 'dart:html';
import '../../../Tools/DescriptionParser.dart';
import 'FullView.dart';

class ThreadDisplay {
  DivElement main;
  DivElement mainView, profileSection, titleSection, descriptionSection;
  SpanElement title, owner;
  ImageElement image;
  DivElement description;
  //Edit stuff
  DivElement editView;
  TextInputElement editTitle;
  TextAreaElement editDescription;
  ButtonElement editButton;
  ButtonElement deleteButton;
  ButtonElement switchToEditViewButton;
  FullView controller;
  bool onMainView = true;
  ThreadDisplay(FullView c) {
    controller = c;
    makeMain();
    makeMainView();
    makeTitleComponent();
    makeImageComponent();
    makeDescriptionComponent();
    makeOwnerComponent();
    makeSwitchToEditViewButton();
    //Make the edit stuff
    makeEditView();
    makeEditTitleComponent();
    makeEditDescriptionComponent();
    makeEditButtonComponent();
    makeDeleteButtonComponent();
    main.append(mainView);
  }
  void updateTitle() {
    title.text = controller.thread.getTitle();
    editTitle.value = controller.thread.getTitle();
  }
  void updateOwner() {
    owner.text = controller.thread.getOwner();
    //If owner is self add edit button
    if (controller.thread.getOwner() == controller.threadController.sectionController.website.database.name) {
      descriptionSection.append(switchToEditViewButton);
    }
  }
  void updateAvatarURL() {
    image.src = controller.thread.getAvatarURL();
    }
  void updateDescription() {
    description.children.clear();
    description.append(DescriptionParser.parseDescription(controller.thread.getDescription()));
    editDescription.text = controller.thread.getDescription();
  }
  DivElement getDiv() {
    return main;
  }
  void saveEdit() {
    //Send new post to server
    Map message = new Map();
    message['Controller'] = controller.threadController.sectionController.name;
    message['Title'] = 'Edit Post';
    message['Thread ID'] = controller.thread.getID();
    message['Edit Title'] = editTitle.value;
    message['Text'] = editDescription.value;
    controller.threadController.sectionController.website.networkingController.Send(message);
    switchView();
  }
  void deleteThread() {
    //Send new post to server
    Map message = new Map();
    message['Controller'] = controller.threadController.sectionController.name;
    message['Title'] = 'Delete Post';
    message['Thread ID'] = controller.thread.getID();
    controller.threadController.sectionController.website.networkingController.Send(message);
    controller.threadController.restoreToDefaultState();
  }
  void switchView() {
    onMainView = !onMainView;
    if (onMainView) {
      switchToMainView();
    } else {
      switchToEditView();
    }
  }
  void switchToMainView() {
    editView.remove();
    main.append(mainView);
  }
  void switchToEditView() {
    mainView.remove();
    main.append(editView);
  }
  //Make GUI Components
  void makeMain() {
    main = new DivElement();
  }
  void makeMainView() {
    mainView = new DivElement();
    mainView.style..display = "inline-block"..marginLeft = "75px";
    profileSection = new DivElement();
    profileSection.style
        ..display = "inline-block"
        ..width = "60px"
        ..position = "absolute"
        ..top = "5px"
        ..textAlign = "center"..color = "white";
    mainView.append(profileSection);
    titleSection = new DivElement();
    titleSection.style
        ..display = "inline-block"
        ..width = "calc(100% - 80px)"
        ..minHeight = "80px"
        ..marginLeft = "70px"
        ..textAlign = "left"..color = "white";
    mainView.append(titleSection);
    descriptionSection = new DivElement();
    descriptionSection.style
            ..display = "inline-block"
            ..width = "calc(100% - 90px)"
            ..marginLeft = "70px"
            ..textAlign = "left"..color = "white";
        mainView.append(descriptionSection);

        mainView.append(new HRElement());
  }
  void makeTitleComponent() {
    title = new SpanElement();
    title.style
        ..fontSize = "50px"
        ..textDecoration = "underline"
        ..paddingTop = "10px"
        ..color = "white";
    titleSection.append(title);
  }
  void makeOwnerComponent() {
    owner = new SpanElement();
    owner.style
        ..fontSize = "15px"
        ..color = "white";
    profileSection.append(owner);
  }
  void makeImageComponent() {
    image = new ImageElement();
    image.style
        ..height = "60px"
        ..width = "60px"
        ..backgroundColor = "rgb(200,200,200)"
        ..borderStyle = "solid"
        ..borderWidth = "1px";
    profileSection.append(image);
  }
  void makeDescriptionComponent() {
    description = new DivElement();
    description.style
        ..wordWrap = "normal"
        //..whiteSpace = "pre"
        ..width = "90%"
        ..color = "white";
    descriptionSection.append(description);
  }
  void makeSwitchToEditViewButton() {
    switchToEditViewButton = new ButtonElement();
    switchToEditViewButton.style
        //..padding = "10px"
        ..color = "black"
        ..float = "right";
    //..width = "100px";
    switchToEditViewButton.text = "Edit";
    switchToEditViewButton.onClick.listen((_) {
      switchView();
    });
  }
  //Edit stuff
  void makeEditView() {
    editView = new DivElement();
    editView.style
        ..margin = "30px"
        ..marginLeft = "70px";
  }
  void makeEditTitleComponent() {
    editTitle = new TextInputElement();
    editTitle.style
        //..fontSize = "20px"
        ..padding = "10px"
        ..color = "black"
        ..width = "calc(100% - 20px)";
    editTitle.placeholder = "Title here";
    editView.append(editTitle);
  }
  void makeEditDescriptionComponent() {
    editDescription = new TextAreaElement();
    editDescription.style
        //..fontSize = "20px"
        ..color = "black"
        ..width = "100%"
        ..height = "100px";
    editDescription.placeholder = "Description here";
    editView.append(editDescription);
  }
  void makeEditButtonComponent() {
    editButton = new ButtonElement();
    editButton.style
        //..fontSize = "20px"
        //..padding = "10px"
        ..color = "black"
        ..float = "right";
    editButton.text = "Save Edit";
    editView.append(editButton);
    editButton.onClick.listen((_) {
      saveEdit();
    });
  }
  void makeDeleteButtonComponent() {
    deleteButton = new ButtonElement();
    deleteButton.style
        //..fontSize = "20px"
        //..padding = "10px"
        ..color = "black"
        ..float = "left";
    deleteButton.text = "Delete Thread";
    editView.append(deleteButton);
    deleteButton.onClick.listen((_) {
      deleteThread();
    });
  }
}

File: ROID-Forum-Web-Client/src/Interface/CodingSection/ThreadViews/ThreadDisplay.ts
import {FullView} from "./FullView";
import {DescriptionParser} from "../../../Utility/DescriptionParser";
import {Utility} from "../../../Utility/Utility";
import {Interface} from "../../../Utility/Interface";

export class ThreadDisplay {
    main: HTMLDivElement;
    mainView: HTMLDivElement;
    profileSection: HTMLDivElement;
    titleSection: HTMLDivElement;
    descriptionSection: HTMLDivElement;
    title: HTMLSpanElement;
    owner: HTMLSpanElement;
    image: HTMLImageElement;
    description: HTMLDivElement;
    //Edit stuff
    editView: HTMLDivElement;
    editTitle: HTMLInputElement;
    editDescription: HTMLTextAreaElement;
    editButton: HTMLButtonElement;
    deleteButton: HTMLButtonElement;
    switchToEditViewButton: HTMLButtonElement;
    controller: FullView;
    onMainView = true;

    constructor(c: FullView) {
        this.controller = c;
        this.makeMain();
        this.makeMainView();
        this.makeTitleComponent();
        this.makeImageComponent();
        this.makeDescriptionComponent();
        this.makeOwnerComponent();
        this.makeSwitchToEditViewButton();
        //Make the edit stuff
        this.makeEditView();
        this.makeEditTitleComponent();
        this.makeEditDescriptionComponent();
        this.makeEditButtonComponent();
        this.makeDeleteButtonComponent();
        this.main.appendChild(this.mainView);
    }

    updateTitle() {
        this.title.innerText = this.controller.thread.getTitle();
        this.editTitle.value = this.controller.thread.getTitle();
    }

    updateOwner() {
        this.owner.innerText = this.controller.thread.getOwner();
        //If this.owner is self add edit button
        if (this.controller.thread.getOwner() == this.controller.threadController.sectionController.website.database.name) {
            this.descriptionSection.appendChild(this.switchToEditViewButton);
        }
    }

    updateAvatarURL() {
        this.image.src = this.controller.thread.getAvatarURL();
    }

    updateDescription() {
        Utility.ClearElements(this.description);
        this.description.appendChild(DescriptionParser.parseDescription(this.controller.thread.getDescription()));
        this.editDescription.innerText = this.controller.thread.getDescription();
    }

    getDiv(): HTMLDivElement {
        return this.main;
    }

    saveEdit() {
        //Send new post to server
        let message: any = {};
        message['Controller'] = this.controller.threadController.sectionController.name;
        message['Title'] = 'Edit Post';
        message['Thread ID'] = this.controller.thread.getID();
        message['Edit Title'] = this.editTitle.value;
        message['Text'] = this.editDescription.value;
        this.controller.threadController.sectionController.website.networkController.send(message);
        this.switchView();
    }

    deleteThread() {
        //Send new post to server
        let message: any = {};
        message['Controller'] = this.controller.threadController.sectionController.name;
        message['Title'] = 'Delete Post';
        message['Thread ID'] = this.controller.thread.getID();
        this.controller.threadController.sectionController.website.networkController.send(message);
        this.controller.threadController.restoreToDefaultState();
    }

    switchView() {
        this.onMainView = !this.onMainView;
        if (this.onMainView) {
            this.switchToMainView();
        } else {
            this.switchToEditView();
        }
    }

    switchToMainView() {
        this.editView.remove();
        this.main.appendChild(this.mainView);
    }

    switchToEditView() {
        this.mainView.remove();
        this.main.appendChild(this.editView);
    }

    //Make GUI Components
    makeMain() {
        this.main = Interface.Create({type: 'div'});
    }

    makeMainView() {
        this.mainView = Interface.Create({type: 'div'});
        this.mainView.style.display = "inline-block";
        this.mainView.style.marginLeft = "75px";
        this.profileSection = Interface.Create({type: 'div'});
        this.profileSection.style.display = "inline-block";
        this.profileSection.style.width = "60px";
        this.profileSection.style.position = "absolute";
        this.profileSection.style.top = "5px";
        this.profileSection.style.textAlign = "center";
        this.profileSection.style.color = "white";
        this.mainView.appendChild(this.profileSection);
        this.titleSection = Interface.Create({type: 'div'});
        this.titleSection.style.display = "inline-block";
        this.titleSection.style.width = "calc(100% - 80px)";
        this.titleSection.style.minHeight = "80px";
        this.titleSection.style.marginLeft = "70px";
        this.titleSection.style.textAlign = "left";
        this.titleSection.style.color = "white";
        this.mainView.appendChild(this.titleSection);
        this.descriptionSection = Interface.Create({type: 'div'});
        this.descriptionSection.style.display = "inline-block";
        this.descriptionSection.style.width = "calc(100% - 90px)";
        this.descriptionSection.style.marginLeft = "70px";
        this.descriptionSection.style.textAlign = "left";
        this.descriptionSection.style.color = "white";
        this.mainView.appendChild(this.descriptionSection);

        this.mainView.appendChild(Interface.Create({type: 'hr'}));
    }

    makeTitleComponent() {
        this.title = Interface.Create({type: 'span'});
        this.title.style.fontSize = "50px";
        this.title.style.textDecoration = "underline";
        this.title.style.paddingTop = "10px";
        this.title.style.color = "white";
        this.titleSection.appendChild(this.title);
    }

    makeOwnerComponent() {
        this.owner = Interface.Create({type: 'span'});
        this.owner.style.fontSize = "15px";
        this.owner.style.color = "white";
        this.profileSection.appendChild(this.owner);
    }

    makeImageComponent() {
        this.image = Interface.Create({type: 'img'});
        this.image.style.height = "60px";
        this.image.style.width = "60px";
        this.image.style.backgroundColor = "rgb(200,200,200)";
        this.image.style.borderStyle = "solid";
        this.image.style.borderWidth = "1px";
        this.profileSection.appendChild(this.image);
    }

    makeDescriptionComponent() {
        this.description = Interface.Create({type: 'div'});
        this.description.style.wordWrap = "normal";
        this.description.style.width = "90%";
        this.description.style.color = "white";
        this.descriptionSection.appendChild(this.description);
    }

    makeSwitchToEditViewButton() {
        this.switchToEditViewButton = Interface.Create({type: 'button'});
        this.switchToEditViewButton.style.color = "black";
        this.switchToEditViewButton.style.cssFloat = "right";
        //this.switchToEditViewButton.style.width = "100px";
        this.switchToEditViewButton.innerText = "Edit";
        this.switchToEditViewButton.onclick = () => {
            this.switchView();
        };
    }

    //Edit stuff
    makeEditView() {
        this.editView = Interface.Create({type: 'div'});
        this.editView.style.margin = "30px";
        this.editView.style.marginLeft = "70px";
    }

    makeEditTitleComponent() {
        this.editTitle = Interface.Create({type: 'input'});
        this.editTitle.style.padding = "10px";
        this.editTitle.style.color = "black";
        this.editTitle.style.width = "calc(100% - 20px)";
        this.editTitle.placeholder = "Title here";
        this.editView.appendChild(this.editTitle);
    }

    makeEditDescriptionComponent() {
        this.editDescription = Interface.Create({type: 'textarea'});
        this.editDescription.style.color = "black";
        this.editDescription.style.width = "100%";
        this.editDescription.style.height = "100px";
        this.editDescription.placeholder = "Description here";
        this.editView.appendChild(this.editDescription);
    }

    makeEditButtonComponent() {
        this.editButton = Interface.Create({type: 'button'});
        this.editButton.style.color = "black";
        this.editButton.style.cssFloat = "right";
        this.editButton.innerText = "Save Edit";
        this.editView.appendChild(this.editButton);
        this.editButton.onclick = () => {
            this.saveEdit();
        };
    }

    makeDeleteButtonComponent() {
        this.deleteButton = Interface.Create({type: 'button'});
        this.deleteButton.style.color = "black";
        this.deleteButton.style.cssFloat = "left";
        this.deleteButton.innerText = "Delete Thread";
        this.editView.appendChild(this.deleteButton);
        this.deleteButton.onclick = () => {
            this.deleteThread();
        };
    }
}